src/Adapter/Presenter/Cart/CartLazyArray.php

```diff
@@ -0,0 +1,604 @@
+                    : $this->translator->trans('Taxes', [], 'Shop.Theme.Checkout'),
+                'amount' => $taxAmount,
+                'value' => $this->priceFormatter->format($taxAmount),
+            ];
+        }
+
+        $this->subTotals = $subtotals;
+
+        return $this->subTotals;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getProductsCount(): int
+    {
+        // If product list is already available, no need to execute a new sql query
+        if (isset($this->products)) {
+            $this->productsCount = array_reduce($this->products, function ($count, $product) {
+                return $count + $product['quantity'];
+            }, 0);
+
+            return $this->productsCount;
+        }
+
+        // Product list is not available, only query the nb of products
+        $this->productsCount = (int) $this->cart->getNbProducts($this->cart->id);
+
+        return $this->productsCount;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getSummaryString(): string
+    {
+        $productsCount = $this->getProductsCount();
+
+        $this->summaryString = $productsCount === 1 ?
+            $this->translator->trans('1 item', [], 'Shop.Theme.Checkout') :
+            $this->translator->trans('%count% items', ['%count%' => $productsCount], 'Shop.Theme.Checkout');
+
+        return $this->summaryString;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getLabels(): array
+    {
+        $this->labels = [
+            'tax_short' => ($this->cartPresenter->includeTaxes())
+                ? $this->translator->trans('(tax incl.)', [], 'Shop.Theme.Global')
+                : $this->translator->trans('(tax excl.)', [], 'Shop.Theme.Global'),
+            'tax_long' => ($this->cartPresenter->includeTaxes())
+                ? $this->translator->trans('(tax included)', [], 'Shop.Theme.Global')
+                : $this->translator->trans('(tax excluded)', [], 'Shop.Theme.Global'),
+        ];
+
+        return $this->labels;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getIdAddressDelivery(): ?int
+    {
+        $this->idAddressDelivery = $this->cart->id_address_delivery;
+
+        return $this->idAddressDelivery;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getIdAddressInvoice(): ?int
+    {
+        $this->idAddressInvoice = $this->cart->id_address_invoice;
+
+        return $this->idAddressInvoice;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getIsVirtual(): bool
+    {
+        $this->isVirtual = $this->cart->isVirtualCart();
+
+        return $this->isVirtual;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getVouchers(): array
+    {
+        $this->vouchers = $this->getTemplateVarVouchers();
+
+        return $this->vouchers;
+    }
+
+    /**
+     * @arrayAccess
+     */
+    public function getDiscounts(): array
+    {
+        $vouchers = $this->getVouchers();
```

Comment:
Wrapping this in `if ($this->discounts === null)` will avoid repeating the logic on another access.

Reply:
then I have to declare $discounts like this: `private array|null $discounts = null;`, which is a bit annoying

otherwise I get an error of the kind `must not be accessed before initialization` 

but in this case it's interesting indeed 👍

Reply:
ah but I forgot, there's no need, the results of all these getters is cached in AbstractLazyArray, it's the point (among other things)

Reply:
maybe all these getters should have protected instead of public accesses 🤔

Reply:
I didn't know this at all... :D

So, when I have been working on product presenters, checks like this are useless? https://github.com/PrestaShop/PrestaShop/blob/a9e361099940601237ba1f554a9f2fba76488cf9/src/Adapter/Presenter/Product/ProductLazyArray.php#L354

Reply:
yep :D  if it used as an array, as it should be

but not if you call the getter directly, but we shouldn't call the getter directly, only `AbstractLazyArray` should call them...  that's why I wonder wether we should put those getters as protected instead...